core/src/main/scala/akka/kafka/scaladsl/Consumer.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.kafka.scaladsl
+
+import scala.concurrent.Future
+import akka.Done
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import akka.kafka.internal.ConsumerStage
+import akka.NotUsed
+import akka.dispatch.ExecutionContexts
+import scala.concurrent.ExecutionContext
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import akka.kafka.internal.ConsumerStage.CommittableOffsetBatchImpl
+import akka.kafka.internal.PlainConsumerStage
+import akka.kafka.internal.CommittableConsumerStage
+
+object Consumer {
+
+  trait Message[K, V] {
+    def key: K
+    def value: V
+    def offset: PartitionOffset
+  }
+
+  final case class CommittableMessage[K, V](
+      key: K,
+      value: V,
+      committableOffset: CommittableOffset
+  ) extends Message[K, V] {
+
+    override def offset: PartitionOffset = committableOffset.offset
+  }
+
+  /**
+   * This interface might move into `akka.stream`
+   */
+  trait Committable {
+    def commit(): Future[Done]
+  }
+
+  trait CommittableOffset extends Committable {
+    def offset: PartitionOffset
+  }
+
+  final case class PartitionOffset(topic: String, partition: Int, offset: Long)
+
+  object CommittableOffsetBatch {
```

Comment:
In the end the commitAsync must be performed by the stage. The stage emits messages that contains individual CommitableOffsetImpl, with ref to their origin stage. The purpose of this CommitableOffsetBatch is to gather several CommittableOffset and commit them in one call to the KafkaKonsumer. Therefore, to commit the batch I need the reference to the stage.

I'm not sure I understand what you mean.